File: exercise/144/morris.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> result;
        TreeNode *now = root;
        while (now) {
            if (!now->left) {
                result.push_back(now->val);
                now = now->right;
                continue;
            }
            TreeNode *pre = now->left;
            while (pre->right && pre->right != now) {
                pre = pre->right;
            }
            if (pre->right != now) {
                pre->right = now;
                result.push_back(now->val);
                now = now->left;
            } else {
                pre->right = NULL;
                now = now->right;
            }
        }
        return result;
    }
};

File: exercise/Binary-Tree/144/morris.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        result = []
        p = root
        while p:
            if p.left is None:
                result.append(p.val)
                p = p.right
            else:
                pre = p.left
                while pre.right and pre.right != p:
                    pre = pre.right
                if pre.right == p:
                    pre.right = None
                    p = p.right
                else:
                    pre.right = p
                    result.append(p.val)
                    p = p.left
        return result

